tests/drivers/counter/counter_basic_api/src/test_counter.c

```diff
@@ -0,0 +1,316 @@
+/*
+ * Copyright (c) 2018, Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <counter.h>
+#include <ztest.h>
+#include <kernel.h>
+
+static volatile int wrap_cnt;
+static volatile int alarm_cnt;
+
+static void wrap_handler(struct device *dev, void *user_data);
+
+void *exp_user_data = (void *)199;
+
+#define WRAP_PERIOD_US 20000
+
+struct counter_alarm_cfg alarm_cfg;
+struct counter_alarm_cfg alarm_cfg2;
+
+const char *devices[] = {
+
+#ifdef CONFIG_COUNTER_TIMER0
+	CONFIG_COUNTER_0_NAME,
+#endif
+#ifdef CONFIG_COUNTER_TIMER1
+	CONFIG_COUNTER_1_NAME,
+#endif
+#ifdef CONFIG_COUNTER_TIMER2
+	CONFIG_COUNTER_2_NAME,
+#endif
+#ifdef CONFIG_COUNTER_TIMER3
+	CONFIG_COUNTER_3_NAME,
+#endif
+#ifdef CONFIG_COUNTER_TIMER4
+	CONFIG_COUNTER_4_NAME,
+#endif
+#ifdef CONFIG_COUNTER_RTC0
+	CONFIG_COUNTER_RTC0_NAME,
+#endif
+	/* RTC0 may be used for system clock */
+#ifdef CONFIG_COUNTER_RTC2
+	CONFIG_COUNTER_RTC2_NAME,
+#endif
+};
+typedef void (*counter_test_func_t)(const char *dev_name);
+
+
+static void counter_tear_down_instance(const char *dev_name)
+{
+	int err;
+	struct device *dev;
+
+	dev = device_get_binding(dev_name);
+
+	err = counter_set_wrap(dev, counter_get_max_wrap(dev), NULL, NULL);
+	zassert_equal(0, err, "Counter wrap to default failed\n");
+
+	err = counter_stop(dev);
+	zassert_equal(0, err, "Counter failed to stop\n");
+}
+
+static void test_all_instances(counter_test_func_t func)
+{
+	for (int i = 0; i < ARRAY_SIZE(devices); i++) {
+		func(devices[i]);
+		counter_tear_down_instance(devices[i]);
+	}
+}
+
+void counter_tear_down(void)
+{
+
+}
+
+static void wrap_handler(struct device *dev, void *user_data)
+{
+	zassert_true(user_data == exp_user_data, "Unexpected callback\n");
+	wrap_cnt++;
+}
+
+#include <nrf_gpio.h>
+void test_wrap_alarm_instance(const char *dev_name)
+{
+	struct device *dev;
+	int err;
+	int cnt;
```

Comment:
@nordic-krch  isn't cnt variable should be u32_t ? In case 0f aonpt it is causing overflow as  aonpt is down counter and intial value of counter 0xffffffff.

Reply:
yes, i've updated that in this PR.